backends/tc/ebpfCodeGen.cpp

```diff
@@ -1456,8 +1464,142 @@ void ControlBodyTranslatorPNA::processFunction(const P4::ExternFunction *functio
+
+                    if (param->is<IR::StructExpression>()) {
+                        auto paramList = action->getParameters();
+                        auto components = param->to<IR::StructExpression>()->components;
+                        if (paramList->parameters.size() != components.size()) {
+                            ::error(ErrorType::ERR_UNEXPECTED,
+                                    "Action params in add_entry should be same as no of action "
+                                    "parameters. %1%",
+                                    action);
+                        }
+                        if (components.size() == 0) {
+                            break;
+                        }
+
+                        builder->emitIndent();
+                        builder->appendLine("struct act_param {");
+                        builder->emitIndent();
+                        int index = 0;
+                        for (auto param : paramList->parameters) {
+                            if (param->direction != IR::Direction::None) {
+                                ::error(ErrorType::ERR_UNEXPECTED,
+                                        "Parameters of action called from add_entry should be "
+                                        "directionless. %1%",
+                                        actionName);
+                            }
+                            auto type = param->type;
+                            auto etype = EBPF::EBPFTypeFactory::instance->create(type);
+                            std::stringstream paramName;
+                            paramName << "param" << index;
+                            builder->append("    ");
+                            etype->declare(builder, paramName.str().c_str(), false);
+                            builder->endOfStatement();
+                            builder->newline();
+                            builder->emitIndent();
+                            index = index + 1;
+                        }
+                        builder->appendLine("};");
+                        builder->emitIndent();
+                        builder->appendLine(
+                            "struct act_param* params = (struct act_param *) "
+                            "update_act_bpf.params;");
+                        // Assign Variables
+                        for (size_t index = 0; index < components.size(); index++) {
+                            std::stringstream paramName;
+                            paramName << "param" << index;
+                            builder->emitIndent();
+                            builder->appendFormat("params->%s = ", paramName.str().c_str());
+                            visit(components.at(index)->expression);
+                            builder->endOfStatement();
+                            builder->newline();
+                        }
+                    }
```

Comment:
This function has gone too big. See if the action related checks can be moved to separate function

Reply:
addressed